app/services/batched_remove_status_service.rb

```diff
@@ -3,32 +3,52 @@
 class BatchedRemoveStatusService < BaseService
   include Redisable
 
-  # Delete given statuses and reblogs of them
-  # Dispatch PuSH updates of the deleted statuses, but only local ones
-  # Dispatch Salmon deletes, unique per domain, of the deleted statuses, but only local ones
-  # Remove statuses from home feeds
-  # Push delete events to streaming API for home feeds and public feeds
-  # @param [Enumerable<Status>] statuses A preferably batched array of statuses
+  # Delete multiple statuses and reblogs of them as efficiently as possible
+  # @param [Enumerable<Status>] statuses An array of statuses
   # @param [Hash] options
-  # @option [Boolean] :skip_side_effects
+  # @option [Boolean] :skip_side_effects Do not modify feeds and send updates to streaming API
   def call(statuses, **options)
-    statuses = Status.where(id: statuses.map(&:id)).includes(:account).flat_map { |status| [status] + status.reblogs.includes(:account).to_a }
+    ActiveRecord::Associations::Preloader.new.preload(statuses, options[:skip_side_effects] ? :reblogs : [:account, reblogs: :account])
 
-    @mentions = statuses.each_with_object({}) { |s, h| h[s.id] = s.active_mentions.includes(:account).to_a }
-    @tags     = statuses.each_with_object({}) { |s, h| h[s.id] = s.tags.pluck(:name) }
+    statuses_and_reblogs = statuses.flat_map { |status| [status] + status.reblogs }
 
-    @json_payloads = statuses.each_with_object({}) { |s, h| h[s.id] = Oj.dump(event: :delete, payload: s.id.to_s) }
+    # Notifications are associated polymorphically and therefore not
+    # in the foreign key system, for this reason, we have to delete them
+    # first. They don't have any side-effects, so delete can be used.
+    Notification.joins(:status).where(status: statuses_and_reblogs).delete_all
+    Notification.joins(favourite: :status).where(favourites: { status: statuses_and_reblogs }).delete_all
+    Notification.joins(mention: :status).where(mentions: { status: statuses_and_reblogs }).delete_all
```

Comment:
Thinking more about it, in the context of account deletion, we want to remove *all* notifications generated by the to-be-deleted account anyway, and we can do that efficiently with something along the lines of `Notification.where(from_account: account).in_batches(&:delete_all)` (or even `Notification.where(from_account: account).delete_all`), which would mean fewer and simpler SQL queries.

Furthermore, this would prevent keeping notifications from reported toots in case the toots themselves are kept around.

Reply:
Actually, I'm somewhat mistaken here, the three types of notifications can be broken into two groups:
- notifications *from* the account being deleted (`mention`, `reblog` and `status`)
- notifications *to* the account being deleted (`reblog`, `favourite`)

the former can safely be deleted as I suggested, the latter would be deleted whenever the account itself would be deleted, through cascading, but we could and maybe should do it explicitly in batches before that